SMART_BATTERY_* messages: Added

hamishwillee opened the pull request:
This implements the messages/enum discussed in #988.

Minor delta since then 
- now have both `capacity_full` and `capacity_full_specification`
- The messages are marked `wip`

@LorenzMeier
- You indicated there might be more faults to [include from the spec](http://sbs-forum.org/specs/sbdat110.pdf)? These can be added as needed to the enum, which has plenty of spare space
- You also suggested `cell_offset`, `battery_id` might be changed to `uint16_t` so that voltage array will be ordered last, and may be truncated if empty. Is this something you still want?

Anyone else got any feedback?

amilcarlucas commented on message_definitions/v1.0/common.xml:
```diff
@@ -2626,6 +2626,21 @@
     </enum>
+    <enum name="MAV_SMART_BATTERY_FAULT">
+      <description>Smart battery supply status/fault flags (bitmask) for health indication.</description>
+      <entry value="0" name="MAV_SMART_BATTERY_FAULT_OK">
```
If this is a bitmask then 0 should not be defined

amilcarlucas commented on message_definitions/v1.0/common.xml:
```diff
@@ -5054,6 +5069,33 @@
     </message>
+    <!-- Smart battery messages -->
+    <message id="370" name="SMART_BATTERY_INFO">
+      <wip/>
```
Please add the same wip comment that other wip messages also have.

hamishwillee commented:
@MaEtUgR,  @apache405, @Naterater, @WickedShell , @AlexKlimaj FYI, I added devcall label, so hopefully this will get final review and merged this week. If there is any particular remaining feedback on this, please let me know in the next day or so.

AlexKlimaj commented:
It's looking great! Glad to see this moving forward. 

My only comment is maybe adding a couple more fault states. One for over current and one under/over temperature.

hamishwillee commented:
Thanks @AlexKlimaj  - let's add them along with any others under discussion following dev meeting.

FYI, Tests failing because grams not yet supported as a unit. Updated in https://github.com/ArduPilot/pymavlink/pull/265

julianoes commented:
How does this relate to `BATTERY_STATUS`? Does it deprecate it or is it an alternative next to it? Or should we extend `BATTERY_STATUS` alternatively?

hamishwillee commented:
> How does this relate to `BATTERY_STATUS`? Does it deprecate it or is it an alternative next to it? Or should we extend `BATTERY_STATUS` alternatively?

Good question. I've been thinking of these as running in parallel for smart batteries. However generally I think this could replace the existing message - @MaEtUgR  @LorenzMeier - what was your thinking?

We can't really extend BATTERY_STATUS - this is two messages.

hamishwillee commented:
Note, 
- Works for both smart battery to flight controller, and from flight controller to ground station. 
- Replacement for BATTERY_STATUS for smart batteries.

amilcarlucas commented on message_definitions/v1.0/common.xml:
```diff
@@ -5057,6 +5069,35 @@
+      <field type="uint8_t" default="0" name="cell_offset">The cell number of the first index in the 'voltages' array field. Using this field allows you to specify cell voltages for batteries with more than 16 cells.</field>
+      <field type="uint16_t[16]" name="voltages" units="mV">Individual cell voltages.  Batteries with more 16 cells can use the cell_offset field to specify the cell offset for the array specified in the current message . Index values above the valid cell count for this battery should have the UINT16_MAX value.</field>
+      <field type="int32_t" name="fault_bitmask" display="bitmask" enum="MAV_SMART_BATTERY_FAULT">Fault/health indications. -1: field not provided.</field>
+      <field type="int32_t" name="time_remaining" units="s">Estimated remaining battery time. -1: field not provided.</field>
```
Would it not make more sense to move the voltages array to the last field in order to let mavlink2 automatically truncate the array ?

hamishwillee replied:
@amilcarlucas This one is more complicated. The ordering is by size first, and only based on declaration order within the common sizes.

So to move the voltages to the end I would need to change `id` and `cell_offset` from `uint8_t` to `uint16_t`. So basically 2 bytes extra for every message. That said, I think it is worth it, because we know that as long as we have less than 14 cells we're saving space.

Seems a reasonable tradeoff (until someone decides to create and extension message that breaks this. I'll change.

hamishwillee replied:
Done

amilcarlucas commented on message_definitions/v1.0/common.xml:
```diff
@@ -5057,6 +5069,35 @@
+      <field type="int32_t" name="capacity_full" units="mAh">Capacity when full (accounting for battery degradation), -1: field not provided.</field>
+      <field type="uint16_t" name="cycle_count">Charge/discharge cycle count. -1: field not provided.</field>
+      <field type="int32_t" name="serial_number">Serial number. -1: field not provided.</field>
+      <field type="char[50]" name="device_name">Static device name. Encode as manufacturer and product names separated using an underscore.</field>
```
Would it not make more sense to move the device_name to the last field in order to let mavlink2 automatically truncate the array ?

hamishwillee replied:
@amilcarlucas No need - reordering automatically does that for us (smallest items are reordered to the end, with items of the same size in the same order - as long as this is below the `id` (which has the same size) it will be ordered last.

hamishwillee commented:
OK, I made some final changes to:
- Add the new faults suggested by @AlexKlimaj 
- Increase size of `SMART_BATTERY_STATUS.id` and `cell_offset` to uint_16 so that battery voltage array can be truncated if needed (cost is +2 bytes per message likely unused, saving up to 15x2 bytes depending on how many cells used.
- Added notes about when used to both this and BATTERY_STATUS (ie use this for smart batteries, and BATTERY_STATUS for non-smart batteries)
- Added note that this is used for both battery to flight stack, and flight stack to GCS. 
- BATTERY_STATUS - added note that this is for flight stack to GCS.

Most of that was approved in dev call. The rest is non-controversial. I am therefore merging.
